lib/bakeAmbientOcclusion.js

```diff
@@ -43,20 +48,323 @@ function bakeAmbientOcclusion(gltf, options) {
+
+        if (options.shaderMode !== 'multiply') {
+            // Check for a diffuse uniform and note if it's a texture or a term
+            var diffuseParameter = technique.parameters.diffuse;
+            if (!defined(diffuseParameter)) {
+                throw new DeveloperError('In stage bakeAmbientOcclusion: Could not find parameter diffuse in technique ' + techniquesIDsToLookAt[i]);
+            }
+
+            var uniforms = technique.uniforms;
+            // fetch the uniform name for the diffuse uniform
+            for (var uniformName in uniforms) {
+                if (uniforms.hasOwnProperty(uniformName)) {
+                    if (uniforms[uniformName] === 'diffuse') {
+                        glslDiffuseColor = uniformName;
+                        break;
+                    }
+                }
+            }
+            if (diffuseParameter.type === WebGLConstants.SAMPLER_2D) {
+                glslNewVaryings += 'varying vec2 v_ambientOcclusionUV; \n';
+                // fetch the UV attribute name so we can pass it through
+                var attributes = technique.attributes;
+                var passUVThrough;
+                for (var attributeName in attributes) {
+                    if (attributes.hasOwnProperty(attributeName)) {
+                        if (attributes[attributeName] === 'texcoord0') {
+                            passUVThrough = 'v_ambientOcclusionUV = ' + attributeName + ';\n';
+                            break;
+                        }
+                    }
+                }
+                if (!defined(passUVThrough)) {
+                    throw new DeveloperError('In stage bakeAmbientOcclusion: Could not find attribute a_texcoord0 in technique ' + techniquesIDsToLookAt[i]);
+                }
+                glslPassThrough += passUVThrough;
+                glslDiffuseColor = 'texture2D(' + glslDiffuseColor + ', v_ambientOcclusionUV)';
+            }
+            glslDiffuseColor += '.rgb';
+        }
+
+        switch(options.shaderMode) {
+            case 'multiply':
+                glslChangeColor = 'gl_FragColor.rgb *= (1.0 - v_ambientOcclusion); \n';
+                break;
+            case 'replace':
+                glslChangeColor = 'gl_FragColor.rgb = (1.0 - v_ambientOcclusion) * ' + glslDiffuseColor + '; \n';
+                break;
+            default: // mix
+                glslChangeColor = 'gl_FragColor.rgb = mix(gl_FragColor.rgb, (1.0 - v_ambientOcclusion) * ' +
+                    glslDiffuseColor + ', ' + (options.blendAmount + 0.0) + '); \n';
+                break;
+        }
+
+        // Replace the shaders in the gltf
+        var program = gltf.programs[technique.program];
+        if (!shadersEdited[program.vertexShader]) {
+            shadersEdited[program.vertexShader] = true;
+            editShader(shaders[program.vertexShader], glslNewAttributes + glslNewVaryings, 'mainBeforeAO', glslPassThrough);
```

Comment:
Here and below, I don't think we can add `glslNewAttributes` and `glslNewVaryings` at the top of the GLSL code.  For example, if the shader has `#version`, it needs to be the first line before comments and whitespace.  Perhaps add these just before the `mainBeforeAO` function.